File: packages/wasmvm/wasmlib/src/sandboxutils.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::*;
use crate::host::*;

pub struct ScSandboxUtils {}

impl ScSandboxUtils {
    // decodes the specified bech32-encoded string value to its original address
    pub fn bech32_decode(&self, bech32: &str) -> ScAddress {
        address_from_bytes(&sandbox(FN_UTILS_BECH32_DECODE, &string_to_bytes(bech32)))
    }

    // encodes the specified address to a bech32-encoded string
    pub fn bech32_encode(&self, addr: &ScAddress) -> String {
        string_from_bytes(&sandbox(FN_UTILS_BECH32_ENCODE, &address_to_bytes(addr)))
    }

    pub fn bls_address_from_pub_key(&self, pub_key: &[u8]) -> ScAddress {
        address_from_bytes(&sandbox(FN_UTILS_BLS_ADDRESS, pub_key))
    }

    pub fn bls_aggregate_signatures(&self, pub_keys: &[&[u8]], sigs: &[&[u8]]) -> Vec<Vec<u8>> {
        let mut enc = WasmEncoder::new();
        uint32_encode(&mut enc, pub_keys.len() as u32);
        for i in 0..pub_keys.len() {
            enc.bytes(pub_keys[i]);
        }
        uint32_encode(&mut enc, sigs.len() as u32);
        for i in 0..sigs.len() {
            enc.bytes(sigs[i]);
        }
        let res = sandbox(FN_UTILS_BLS_AGGREGATE, &enc.buf());
        let mut dec = WasmDecoder::new(&res);
        return [dec.bytes(), dec.bytes()].to_vec();
    }

    pub fn bls_valid_signature(&self, data: &[u8], pub_key: &[u8], signature: &[u8]) -> bool {
        let mut enc = WasmEncoder::new();
        enc.bytes(data);
        enc.bytes(pub_key);
        enc.bytes(signature);
        bool_from_bytes(&sandbox(FN_UTILS_BLS_VALID, &enc.buf()))
    }

    pub fn ed25519_address_from_pub_key(&self, pub_key: &[u8]) -> ScAddress {
        address_from_bytes(&sandbox(FN_UTILS_ED25519_ADDRESS, pub_key))
    }

    pub fn ed25519_valid_signature(&self, data: &[u8], pub_key: &[u8], signature: &[u8]) -> bool {
        let mut enc = WasmEncoder::new();
        enc.bytes(data);
        enc.bytes(pub_key);
        enc.bytes(signature);
        bool_from_bytes(&sandbox(FN_UTILS_ED25519_VALID, &enc.buf()))
    }

    // hashes the specified value bytes using blake2b hashing and returns the resulting 32-byte hash
    pub fn hash_blake2b(&self, value: &[u8]) -> ScHash {
        hash_from_bytes(&sandbox(FN_UTILS_HASH_BLAKE2B, value))
    }

    // hashes the specified value bytes using blake2b hashing and returns the resulting 32-byte hash
    pub fn hash_name(&self, name: &str) -> ScHname {
        hname_from_bytes(&sandbox(FN_UTILS_HASH_NAME, &string_to_bytes(name)))
    }

    // hashes the specified value bytes using sha3 hashing and returns the resulting 32-byte hash
    pub fn hash_sha3(&self, value: &[u8]) -> ScHash {
        hash_from_bytes(&sandbox(FN_UTILS_HASH_SHA3, value))
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/sandboxutils.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {sandbox} from "./host";
import {
    FnUtilsBech32Decode,
    FnUtilsBech32Encode,
    FnUtilsBlsAddress,
    FnUtilsBlsAggregate,
    FnUtilsBlsValid,
    FnUtilsEd25519Address,
    FnUtilsEd25519Valid,
    FnUtilsHashBlake2b,
    FnUtilsHashName,
    FnUtilsHashSha3
} from "./sandbox";
import {boolFromBytes} from "./wasmtypes/scbool";
import {hashFromBytes, ScHash} from "./wasmtypes/schash";
import {WasmDecoder, WasmEncoder} from "./wasmtypes/codec";
import {addressFromBytes, addressToBytes, ScAddress} from "./wasmtypes/scaddress";
import {stringFromBytes, stringToBytes} from "./wasmtypes/scstring";
import {uint32Encode} from "./wasmtypes/scuint32";
import {hnameFromBytes, ScHname} from "./wasmtypes/schname";

export class ScSandboxUtils {
    // decodes the specified bech32-encoded string value to its original bytes
    public bech32Decode(value: string): ScAddress {
        return addressFromBytes(sandbox(FnUtilsBech32Decode, stringToBytes(value)));
    }

    // encodes the specified bytes to a bech32-encoded string
    public bech32Encode(addr: ScAddress): string {
        return stringFromBytes(sandbox(FnUtilsBech32Encode, addressToBytes(addr)));
    }

    public blsAddressFromPubKey(pubKey: Uint8Array): ScAddress {
        return addressFromBytes(sandbox(FnUtilsBlsAddress, pubKey));
    }

    public blsAggregateSignatures(pubKeys: Uint8Array[], sigs: Uint8Array[]): Uint8Array[] {
        const enc = new WasmEncoder();
        uint32Encode(enc, pubKeys.length as u32);
        for (let i = 0; i < pubKeys.length; i++) {
            enc.bytes(pubKeys[i]);
        }
        uint32Encode(enc, sigs.length as u32);
        for (let i = 0; i < sigs.length; i++) {
            enc.bytes(sigs[i]);
        }
        const result = sandbox(FnUtilsBlsAggregate, enc.buf());
        const decode = new WasmDecoder(result);
        return [decode.bytes(), decode.bytes()];
    }

    public blsValidSignature(data: Uint8Array, pubKey: Uint8Array, signature: Uint8Array): bool {
        const enc = new WasmEncoder().bytes(data).bytes(pubKey).bytes(signature);
        return boolFromBytes(sandbox(FnUtilsBlsValid, enc.buf()));
    }

    public ed25519AddressFromPubKey(pubKey: Uint8Array): ScAddress {
        return addressFromBytes(sandbox(FnUtilsEd25519Address, pubKey));
    }

    public ed25519ValidSignature(data: Uint8Array, pubKey: Uint8Array, signature: Uint8Array): bool {
        const enc = new WasmEncoder().bytes(data).bytes(pubKey).bytes(signature);
        return boolFromBytes(sandbox(FnUtilsEd25519Valid, enc.buf()));
    }

    // hashes the specified value bytes using blake2b hashing and returns the resulting 32-byte hash
    public hashBlake2b(value: Uint8Array): ScHash {
        return hashFromBytes(sandbox(FnUtilsHashBlake2b, value));
    }

    // hashes the specified value bytes using sha3 hashing and returns the resulting 32-byte hash
    public hashSha3(value: Uint8Array): ScHash {
        return hashFromBytes(sandbox(FnUtilsHashSha3, value));
    }

    // hashes the specified value bytes using blake2b hashing and returns the resulting 32-byte hash
    public hname(value: string): ScHname {
        return hnameFromBytes(sandbox(FnUtilsHashName, stringToBytes(value)));
    }
}
